src/Sarif/Writers/SarifLogger.cs

```diff
@@ -241,14 +243,24 @@ public void AnalysisStopped(RuntimeConditions runtimeConditions)
 
         public void Log(IRule rule, Result result)
         {
+            if (rule.Id != result.RuleId)
+            {
+                //The rule id '{0}' specified by the result does not match the actual id of the rule '{1}'
+                string message = string.Format(SdkResources.ResultRuleIdDoesNotMatchRule,
+                    result.RuleId.ToString(),
+                    rule.Id.ToString());
+
+                throw new ArgumentException(message);
+            }
+
             if (!ShouldLog(result.Level))
             {
                 return;
             }
 
             if (rule != null)
             {
-                Rules[rule.Id] = rule;
+                Rules[result.RuleKey ?? result.RuleId] = rule;
```

Comment:
My preference would be to throw if a rule is listed once with its id and one or more times by keys. (Multiple keys are fine, of course, that's why `ruleKey` exists.)